service/header/sync.go

```diff
@@ -20,147 +23,263 @@ type Syncer struct {
-func (s *Syncer) IsSyncing() bool {
-	return atomic.LoadUint64(&s.inProgress) == 1
+// wantSync will trigger the syncing loop (non-blocking).
+func (s *Syncer) wantSync() {
+	select {
+	case s.triggerSync <- struct{}{}:
+	default:
+	}
 }
 
-// syncInProgress indicates Syncer's sync status is in progress.
-func (s *Syncer) syncInProgress() {
-	atomic.StoreUint64(&s.inProgress, 1)
+// syncLoop controls syncing process.
+func (s *Syncer) syncLoop(ctx context.Context) {
+	for {
+		select {
+		case <-s.triggerSync:
+			s.sync(ctx)
+		case <-ctx.Done():
+			return
+		}
+	}
 }
 
-// finishSync indicates Syncer's sync status as no longer in progress.
-func (s *Syncer) finishSync() {
-	atomic.StoreUint64(&s.inProgress, 0)
+// sync ensures we are synced up to any trusted header.
+func (s *Syncer) sync(ctx context.Context) {
+	// indicate syncing
+	atomic.StoreUint64(&s.inProgress, 1)
+	// indicate syncing is stopped
+	defer atomic.StoreUint64(&s.inProgress, 0)
+
+	trstHead, err := s.trustedHead(ctx)
+	if err != nil {
+		log.Errorw("getting trusted head", "err", err)
+		return
+	}
+
+	err = s.syncTo(ctx, trstHead)
+	if err != nil {
+		log.Errorw("syncing headers", "err", err)
+		return
+	}
 }
 
-// Validate implements validation of incoming Headers and stores them if they are good.
-func (s *Syncer) Validate(ctx context.Context, p peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
-	header, err := UnmarshalExtendedHeader(msg.Data)
+// validateMsg implements validation of incoming Headers as PubSub msg validator.
+func (s *Syncer) validateMsg(ctx context.Context, p peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
+	maybeHead, err := UnmarshalExtendedHeader(msg.Data)
 	if err != nil {
-		log.Errorw("unmarshalling ExtendedHeader received from the PubSub",
-			"err", err, "peer", p.ShortString())
+		log.Errorw("unmarshalling header",
+			"from", p.ShortString(),
+			"err", err)
 		return pubsub.ValidationReject
 	}
 
-	// if syncing is still in progress - just ignore the new header as
-	// Syncer will fetch it after anyway, but if syncer is done, append
-	// the header.
-	if !s.IsSyncing() {
-		err := s.store.Append(ctx, header)
-		if err != nil {
-			log.Errorw("appending store with header from PubSub",
-				"hash", header.Hash().String(), "height", header.Height, "peer", p.ShortString())
-			// TODO(@Wondertan): We need to be sure that the error is actually validation error.
-			//  Rejecting msgs due to storage error is not good, but for now that's fine.
+	return s.incoming(ctx, p, maybeHead)
+}
+
+// incoming process new incoming Headers, validates them and applies/caches if applicable.
+func (s *Syncer) incoming(ctx context.Context, p peer.ID, maybeHead *ExtendedHeader) pubsub.ValidationResult {
+	// 1. Try to append. If header is not adjacent/from future - try it for pending cache below
+	err := s.store.Append(ctx, maybeHead)
+	if err == nil {
+		return pubsub.ValidationAccept
+	} else if err != ErrNonAdjacent {
```

Comment:
switch on err?

Reply:
Here it works. Makes code better. Thanks for noticing this.